consensus/polybft/checkpoint_manager.go

```diff
@@ -0,0 +1,209 @@
+	txn := &ethgo.Transaction{
+		To: &checkpointManagerAddr,
+	}
+
+	// detect any pending (previously failed) checkpoints and send them
+	for blockNum := checkpointID + 1; blockNum < latestHeader.Number; blockNum++ {
+		header, exists := c.blockchain.GetHeaderByNumber(blockNum)
+		if !exists {
+			return fmt.Errorf("block %d was not found", blockNum)
+		}
+
+		extra, err := GetIbftExtra(header.ExtraData)
+		if err != nil {
+			return err
+		}
+
+		// send checkpoints only for epoch ending blocks (validators delta is non-empty)
+		if extra.Validators.IsEmpty() {
+			continue
+		}
+
+		err = c.submitCheckpointInternal(nonce, txn, *header, *extra)
+		if err != nil {
+			return err
+		}
+		nonce++
+	}
+
+	extra, err := GetIbftExtra(latestHeader.ExtraData)
+	if err != nil {
+		return err
+	}
+
+	return c.submitCheckpointInternal(nonce, txn, latestHeader, *extra)
+}
+
+// submitCheckpointInternal encodes checkpoint data for the given block and
+// sends a transaction to the CheckpointManager rootchain contract
+func (c *checkpointManager) submitCheckpointInternal(nonce uint64, txn *ethgo.Transaction,
+	header types.Header, extra Extra) error {
+	nextEpochValidators, err := c.consensusBackend.GetValidators(header.Number, nil)
+	if err != nil {
+		return err
+	}
+
+	input, err := c.abiEncodeCheckpointBlock(header, extra, nextEpochValidators)
+	if err != nil {
+		return fmt.Errorf("failed to encode checkpoint data to ABI for block %d: %w", header.Number, err)
+	}
+
+	txn.Input = input
+
+	receipt, err := c.rootchain.SendTransaction(nonce, txn)
+	if err != nil {
+		return err
+	}
+
+	if receipt.Status == uint64(types.ReceiptFailed) {
+		return fmt.Errorf("transaction execution failed for block %d", header.Number)
+	}
+
+	return nil
+}
+
+// abiEncodeCheckpointBlock encodes checkpoint data into ABI format for a given header
+func (c *checkpointManager) abiEncodeCheckpointBlock(header types.Header, extra Extra,
+	nextValidators AccountSet) ([]byte, error) {
+	params := map[string]interface{}{
+		"chainID":             new(big.Int).SetUint64(c.blockchain.GetChainID()),
+		"aggregatedSignature": extra.Committed.AggregatedSignature,
+		"validatorsBitmap":    extra.Committed.Bitmap,
+		"epochNumber":         new(big.Int).SetUint64(extra.Checkpoint.EpochNumber),
+		"blockNumber":         new(big.Int).SetUint64(header.Number),
+		"blockHash":           header.Hash,
+		"blockRound":          new(big.Int).SetUint64(extra.Checkpoint.BlockRound),
+		"eventRoot":           extra.Checkpoint.EventRoot.Bytes(),
+		"nextValidators":      nextValidators.AsGenericMaps(),
+	}
+
+	return submitCheckpointMethod.Encode(params)
+}
+
+// isCheckpointIntervalElapsed returns indication whether given block is the checkpoint block.
+// Returns true at each N blocks, where N is calculated
+// as division between predefined checkpoint interval and block time
+func (c *checkpointManager) isCheckpointIntervalElapsed(blockNumber uint64) bool {
```

Comment:
i think it's not correct, because it can be changed(or terminated earlier).

Reply:
We have agreed that this can be changed through on-chain governance, but for know `checkpointTimeInterval` is just a hardcoded const value (https://github.com/0xPolygon/polygon-edge/pull/861/files#diff-fce12e7457acbf9d31265d9e0f71a4340b87f09cb3dd52e2f61df69d0bbc142dR28).

Reply:
Fixed [732dd70](https://github.com/0xPolygon/polygon-edge/pull/861/commits/732dd70ad34721da848f89c8016ac39031ac3405). Added setter function for setting checkpoints offset.